### ISSUE TITLE
Safari question: «Increase database size?»

### ISSUE THREAD

**Issue by jsfan3** [codenameone/CodenameOne#2699]
After doing some tests with photo capture, Safari gave me this message:
`Increase database size? Do you want to allow "https://..." to use up to 10 MB storage?`
I tapped "Allow", but this popup frozed my web-app, and I needed to reload it. I've done other tests after clearing Safari data, the same question is shown after taking some photos, even if I delete (in the code) the temp files of the photo. This popup always frozes the app, regardless if I tap "Cancel" or "Allow".

I suspect that the problem is that the browser cache is not cleared even if I cancel the temp photos in the code. 

Moreover, my web-app will also use acquiring of videos, so I expect that I'll need a lot more than 10MB of storage.

What do you think about this issue? What does Safari mean for "database size"?

**Comment by shannah** [comment]
It is likely that this will be fixed by simply updating the version of localforage that we are using.  We are on 1.4.3, but in [1.5.0](https://github.com/localForage/localForage/releases/tag/1.5.0) they added IndexedDb support in Safari.

I'll need to experiment

**Comment by jsfan3** [comment]
Now Safari asked me to increase the database size to 25MB...
Why does this request appear? I don't use any database inside the app.

**Comment by shannah** [comment]
We use localForage for FileSystemStorage and Storage.  It provides an abstraction over the local storage APIs provided by the browser, using the "best" option available in that browser.  In Safari, that was WebSQL, in most other's it's IndexedDB.

**Comment by jsfan3** [comment]
Thank you for the clarification :+1:

**Comment by shannah** [comment]
Can you post a test case for this?  I'd like to reproduce it before trying to upgrade the localforage version so I know if that fixes it.

**Comment by jsfan3** [comment]
Yes, of course.

The following test case stores 512MB in the FileSystemStorage and 512MB in the Storage.
It runs fine on Simulator, iOS native app (tested on iPhone X) and Android native app (tested on Android 7), as in the following screenshot:

![testcasestorage](https://user-images.githubusercontent.com/1997316/52598431-fb9d5200-2e55-11e9-9295-3cdd92418c35.png)

**About the Javascript port, it seems to run _(very very slow!!!)_ on Android Chrome and on Linux Desktop Chrome. On my iPhone, it fails on Safari, on Firefox and on Chrome. It fails also on Android Firefox.**

In this test case I use small files of 1MB that can simulate photos. You can change the code and test a lot bigger files (like 10MB or 100MB or even bigger) to simulate the saving of videos. I have to do a social app like Instagram with a lot of photos and videos.

```
    public void start() {
        if (current != null) {
            current.show();
            return;
        }

        // Choose in the following value how many MB you want to try to save.
        // Both FileSystemStorage and Storage are tested.
        // Don't worry: all the written files are deleted at the end of the test
        int testSize = 512; // MB

        Form hi = new Form("Test Case", BoxLayout.y());
        Label savedToFileSystemStorage = new Label("Saved to FileSystemStorage: 0 MB");
        Label savedToStorage = new Label("Saved to Storage: 0 MB");
        hi.addAll(new SpanLabel("Test Case Storage and FileSystemStorage..."),
                savedToFileSystemStorage, savedToStorage);
        hi.show();

        // I run the following test in a separate thread otherwise the UI is not shown until the test is completed
        EasyThread.start("TestFileWriting").run(() -> {
            int fileName = 0;
            byte[] oneMegaByte = new byte[1048576];
            long[] countFileSystemStorage = {0};
            long[] countStorage = {0};

            while (countFileSystemStorage[0] / 1048576 <= testSize && countStorage[0] / 1048576 <= testSize) {
                try {
                    String fileToFS = FileSystemStorage.getInstance().getAppHomePath() + ++fileName + ".raw";
                    String fileToStorage = ++fileName + ".raw";
                    OutputStream fileSystemStorageOut = FileSystemStorage.getInstance().openOutputStream(fileToFS);
                    OutputStream storageOut = Storage.getInstance().createOutputStream(fileToStorage);

                    fileSystemStorageOut.write(oneMegaByte);
                    fileSystemStorageOut.close();
                    storageOut.write(oneMegaByte);
                    storageOut.close();

                    countFileSystemStorage[0] += FileSystemStorage.getInstance().getLength(fileToFS);
                    countStorage[0] += Storage.getInstance().entrySize(fileToStorage);
                    Display.getInstance().callSeriallyAndWait(() -> {
                        savedToFileSystemStorage.setText("Saved to FileSystemStorage: " + countFileSystemStorage[0] / 1024 / 1024 + " MB");
                        savedToStorage.setText("Saved to Storage: " + countStorage[0] / 1024 / 1024 + " MB");
                    });
                } catch (IOException ex) {
                    Log.e(ex);
                }
            }

            // don't forget to clear the memory
            for (int i = 1; i <= fileName; i++) {
                if (Storage.getInstance().exists(i + ".raw")) {
                    Storage.getInstance().deleteStorageFile(i + ".raw");
                } else if (FileSystemStorage.getInstance().exists(i + ".raw")) {
                    FileSystemStorage.getInstance().delete(i + ".raw");
                }
            }
            Display.getInstance().callSeriallyAndWait(() -> {
                hi.add(new Label("Test Finished - All occupied memory is now cleared"));
                hi.revalidate();
            });
        });
    }
```

**Comment by shannah** [comment]
Thanks the test case.
I confirmed that in Safari on Desktop, it periodically shows a dialog prompt asking to increase the database size.  The "insert" that was pending, and prompted the dialog then fails.
Additionally, I confirmed that in Safari on mobile, this test seems to crash at around 250MB saved to storage.

I upgraded to localforage 1.7.3 (from 1.4.7), and it seems to have corrected both issues.  Safari on desktop no longer prompts to increase the database size.  And safari on iOS can now run the test to completion without crashing.

This will be included in the next server update.
